stateresolution.go

```diff
@@ -0,0 +1,255 @@
+package gomatrixserverlib
+
+import (
+	"bytes"
+	"crypto/sha1"
+	"sort"
+)
+
+// ResolveStateConflicts takes a list of state events with conflicting state keys
+// and works out which event should be used for each state event.
+func ResolveStateConflicts(conflicted []Event, authEvents []Event) []Event {
+	var r stateResolver
+	// Group the conflicted events by type and state key.
+	r.addConflicted(conflicted)
+	// Add the unconflicted auth events needed for auth checks.
+	for i := range authEvents {
+		r.addAuthEvent(&authEvents[i])
+	}
+	// Resolve the conflicted auth events.
+	r.resolveAndAddAuthBlocks([][]Event{r.creates})
+	r.resolveAndAddAuthBlocks([][]Event{r.powerLevels})
+	r.resolveAndAddAuthBlocks([][]Event{r.joinRules})
+	r.resolveAndAddAuthBlocks(r.thirdPartyInvites)
+	r.resolveAndAddAuthBlocks(r.members)
+	// Resolve any other conflicted state events.
+	for _, block := range r.others {
+		if event := r.resolveNormalBlock(block); event != nil {
+			r.result = append(r.result, *event)
+		}
+	}
+	return r.result
+}
+
+// A stateResolver tracks the internal state of the state resolution algorithm
+// It has 3 sections:
+//
+//  * Lists of lists of events to resolve grouped by event type and state key.
+//  * The resolved auth events grouped by type and state key.
+//  * A List of resolved events.
+//
+// It implements the AuthEvents interface and can be used for running auth checks.
+type stateResolver struct {
+	creates                   []Event
+	powerLevels               []Event
+	joinRules                 []Event
+	thirdPartyInvites         [][]Event
+	members                   [][]Event
+	others                    [][]Event
+	resolvedCreate            *Event
+	resolvedPowerLevels       *Event
+	resolvedJoinRules         *Event
+	resolvedThirdPartyInvites map[string]*Event
+	resolvedMembers           map[string]*Event
+	result                    []Event
+}
+
+func (r *stateResolver) Create() (*Event, error) {
+	return r.resolvedCreate, nil
+}
+
+func (r *stateResolver) PowerLevels() (*Event, error) {
+	return r.resolvedPowerLevels, nil
+}
+
+func (r *stateResolver) JoinRules() (*Event, error) {
+	return r.resolvedJoinRules, nil
+}
+
+func (r *stateResolver) ThirdPartyInvite(key string) (*Event, error) {
+	return r.resolvedThirdPartyInvites[key], nil
+}
+
+func (r *stateResolver) Member(key string) (*Event, error) {
+	return r.resolvedMembers[key], nil
+}
+
+func (r *stateResolver) addConflicted(events []Event) {
+	type conflictKey struct {
+		eventType string
+		stateKey  string
+	}
+	offsets := map[conflictKey]int{}
+	// Split up the conflicted events into blocks with the same state key.
```

Comment:
I assume by "block" you mean "list"? And I assume by "the same state key" you mean "the same event type and state key"?

Superficially it sounds like you're taking a list of events and putting them into a dict keyed off the `conflictKey` (type/state key tuple), but I must be missing something because you do Magic depending on the event type. Needs much more explanation for me to grok :(